File: LeetCode_Algorithm/383(Ransom Note)/383.py
import collections
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        # sol1 hashMap O(n) 109ms 29%
        if len(ransomNote) > len(magazine): return False

        def makeMap(string):
            temp = {}
            for char in string:
                if char in temp:
                    temp[char] += 1
                else:
                    temp[char] = 1
            return temp

        ransomNote_map = makeMap(ransomNote)
        magazine_map = makeMap(magazine)

        for char in ransomNote_map:
            if char in magazine_map: 
                if magazine_map[char] < ransomNote_map[char]:
                    return False
            else:
                return False
        return True


        
        pass

print(Solution().canConstruct("bg", "efjbdfbdgfjhhaiigfhbaejahgfbbgbjagbddfgdiaigdadhcfcj"))
print(Solution().canConstruct("a", "b"))

File: LeetCode_Algorithm/383(Ransom Note)/383.js
var makeMap = function(note) {
  // 181ms 13%
  let hashMap = new Map();
  for (let char of note) {
    if (hashMap[char] == undefined) {
      hashMap[char] = 1
    } else {
      hashMap[char] += 1
    }
  }
  return hashMap
}
var canConstruct = function(ransomNote, magazine) {
  let ransomNoteMap = makeMap(ransomNote)
  let magazineMap = makeMap(magazine)
  for (let key in ransomNoteMap) {
    if (magazineMap[key] < ransomNoteMap[key]) {
      return false
    } else if (magazineMap[key] == undefined) {
      return false
    }
  }
  return true
};

console.log(canConstruct("aa", "ab"))